components/raftstore-v2/src/operation/command/write/mod.rs

```diff
@@ -65,35 +73,38 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
             }
             Err(req) => {
                 let res = self.propose_command(ctx, req);
-                self.post_propose_write(ctx, res, vec![ch]);
+                self.post_propose_command(ctx, res, vec![ch], call_proposed_on_success);
             }
         }
     }
 
-    #[inline]
-    pub fn post_propose_write<T>(
-        &mut self,
-        ctx: &mut StoreContext<EK, ER, T>,
-        res: Result<u64>,
-        ch: Vec<CmdResChannel>,
-    ) {
-        let idx = match res {
-            Ok(i) => i,
-            Err(e) => {
-                ch.report_error(cmd_resp::err_resp(e, self.term()));
-                return;
-            }
-        };
-        let p = Proposal::new(idx, self.term(), ch);
-        self.enqueue_pending_proposal(ctx, p);
-        self.set_has_ready();
-    }
-
-    pub fn propose_pending_command<T>(&mut self, ctx: &mut StoreContext<EK, ER, T>) {
+    pub fn propose_pending_writes<T>(&mut self, ctx: &mut StoreContext<EK, ER, T>) {
         if let Some(encoder) = self.simple_write_encoder_mut().take() {
+            let call_proposed_on_success = if encoder.notify_proposed() {
+                // The request has pass conflict check and called all proposed callbacks.
+                false
+            } else {
+                // Epoch may have changed since last check.
+                let from_epoch = encoder.header().get_region_epoch();
+                let res = util::compare_region_epoch(
+                    from_epoch,
+                    self.region(),
+                    NORMAL_REQ_CHECK_CONF_VER,
+                    NORMAL_REQ_CHECK_VER,
+                    true,
+                );
+                if let Err(mut e) = res {
+                    // TODO: query sibling regions.
+                    ctx.raft_metrics.invalid_proposal.epoch_not_match.inc();
+                    encoder.encode().1.report_error(cmd_resp::new_error(e));
+                    return;
+                }
+                // Only when it applies to current term, the epoch check can be reliable.
+                self.applied_to_current_term()
+            };
             let (data, chs) = encoder.encode();
```

Comment:
Can we move `not leader` check before here? Then we can save time of encoding.

Reply:
and do lazy encode until here?

Reply:
I don't understand. If it's not leader, then the proposal is rejected in the very beginning. For most cases, if it passes all the check, it's likely to be proposed, lazy encoding won't have much benefits but only adds more memory consumption and make code complicated.